File: 0872-leaf-similar-trees/0872-leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void helper(List<Integer>t1,TreeNode root){
        if(root == null)return;
        if(root.right == null && root.left == null)t1.add(root.val);

        helper(t1,root.left);
        helper(t1,root.right);
    }

    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer>t1 = new ArrayList<>();
        List<Integer>t2 = new ArrayList<>();

        helper(t1,root1);
        helper(t2,root2);

        return t1.equals(t2);
    }
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

 void helper(vector<int>&temp,TreeNode* root){
        if(root == NULL)return;
        if(root -> left == NULL && root -> right == NULL)temp.push_back(root->val);

        helper(temp, root -> left);
        helper(temp, root -> right);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int>temp1;
        vector<int>temp2;

        helper(temp1,root1);
        helper(temp2,root2);

        if(temp1.size() != temp2.size())return false;
        int n = temp1.size();
        for(int i=0;i<n;i++){
            if(temp1[i] != temp2[i])return false;
        }
        return true;
    }
};